### PR TITLE
Add changeset execution time to output

### LINKED ISSUE #5577: Feature Request: Add changeset execution time to database changelog table
### Search first

- [X] I searched and no similar issues were found

### Description

I believe it would be beneficial to developers to be able to identify changesets that take a long time to run. To this end, I think a new column should be added to the database change log table that is execution_milliseconds.

In looking at liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java, a startTime and stopTime are already grabbed and used to display the execution time in milliseconds in the log.

So this change would be minimal in just exposing that data through the class and logging it in the database changeset tabl (and adding a column to the table to hold the information).

I am happy to do a PR, just wanted to make sure this sounded acceptable to the team before I did the actual coding!

### Steps To Reproduce

N/A, feature request

### Expected/Desired Behavior

Described above

### Liquibase Version

_No response_

### Database Vendor & Version

_No response_

### Liquibase Integration

_No response_

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

_No response_

### Additional Context

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR (Thank you!)

### PR DESCRIPTION
**Pull request by jasonlyle88**
## Impact

- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [X] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 

## Description

Fixes #5577 

Changeset execution time is already being calculated to be shown on log output. This PR takes that and also stores the execution time in the database changelog table.

This is accomplished by adding a new column, `EXECUTION_MILLISECONDS` to the database changelog table. This is handled by both the create table statement (`CreateDatabaseChangeLogTableGenerator` class or `CreateDatabaseChangeLogTableGeneratorSybase` class) for when the database changelog table does not exist yet and the update table statement (`StandardChangeLogHistoryService` class) for when the database changelog table already exists.

The `ChangeSet` class was then updated to store the start execution time and end execution time as an attribute of the class and a public method added to return the millisecond difference between the end and start execution times.

The `MarkChangeSetRanGenerator` class then makes sure to either insert or update the `EXECUTION_MILLISECONDS` column when marking the changeset ran and uses the new procedure from the `ChangeSet` class to get the number of milliseconds the changeset took to execute.

## Things to be aware of

Previously, the `ChangeSet` class was capturing the start and end time using the java.time.Date class. The new class, java.time.Instant is preferred for this, and was already partially being used for the MDC Value. As such, I did away with the java.time.Date references and kept everything as a java.time.Instant instance and just converted that to epoch-milliseconds when needed.

I also added two directories to the `.gitignore` file because they were created when I ran `mvn clean install` when there was no differences between my branch and the master branch. If this is unwanted, I can easily remove the `.gitignore` entries... but I figured it made sense to add them. Let me know what to do in this regards.

## Things to worry about

I believe I updated all the code that handles creating/updating the database changelog table, but am not 100% positive. Would love someone familiar with this section of the code to verify this!

## Additional Context

I updated the one existing test that broke based on my changes. This existing test broke because it parses the SQL for updating the database changelog table after a changeset executes by splitting across comma seperated fields. Since I added a new column, the number of the desired field for the test needed to change. That is all I changed on the existing test.

I have not added any tests for the functionality that I wrote. In looking for other tests that deal with setting up the database changelog table, I could not find any, so I'm just keeping with that. I am not sure where to start tests for this if tests need to be added as no such tests currently exist (or I just couldn't find them... which is entirely possible).

### CODE REVIEW

**Comment by jasonlyle88** [comment]
@MalloD12 

That is disappointing to hear. I understand that the `databachangelog` table growing out of control would make for a difficult maintenance situation. And that makes decisions around modifying it important. However, I'm surprised with a feature that has use for clients (hence a PR being created for it) and even potential use for your team internally (as pointed out by you yourself), that the decision was made not to include this. Especially considering the table currently holds all kinds of information about the execution of a changeset (its identifier, what tags were used to run it, when it was run, its deployment id, etc). This information is right in line with all the other information in the table and, honestly, seems like information that should have been included in the first place.

Is there any chance of this being reconsidered?

**Comment by jasonlyle88** [comment]
@MalloD12 I understand the generic worry, but I don't understand the worry in this specific case. Again, this seems like the exact information you want in that table. In fact, looking at the history of the `liquibase-standard/src/main/java/liquibase/sqlgenerator/core/CreateDatabaseChangeLogTableGenerator.java` file, it has in fact been updated over Liquibase's lifetime to add missing changeset execution data. Notably, the `CONTEXTS`, `LABELS`, and even `DEPLOYMENT_ID` were not originally on this table. How does the changeset execution time differ?

However, it should 100% be a part of the MDC logging, I definitely missed that in my current implmentation. But that isn't really that great of a compromise. Isn't the structured logging only available to PRO customers? Seems like not the best look to hide such a basic metric, such as changeset execution time, behind a paywall.

Even if I don't agree with the decision, if I can't change the team's minds, I agree with this being included without the `databasechangelog` modifications at the very least. That enables an extremely small extension to be built that just uses the public function I've already included in this PR to get the execution time from the changeset instance.

**Comment by MalloD12** [comment]
Hi @jasonlyle88.

Our team has discussed this more, and if you are ok we would like to modify this PR a bit so that changeset execution time is displayed in output logs, but not stored in the `databasechangelogtable`. If others would like to store that or use it, they can develop a small extension like you pointed out using the public method you have in the PR. 

If you are ok with that, please move forward with it.

Thanks,
Daniel.

**Comment by jasonlyle88** [comment]
Hey @MalloD12 

I'm glad to hear this! Yeah, I think that sounds like a great to me. I'm just curious what you mean by "output logs"? MDC logging makes sense, but I think if this information is not being stored in the `databasechangelog` table, it would be useful to display it on the execution output (maybe this is what you already mean!). Example:

current execution output:
```
Running Changeset: changelog.xml::my_example_1::jlyle
Running Changeset: changelog.xml::my_example_2::jlyle
Running Changeset: changelog.xml::my_example_3::jlyle
```

proposed execution output:
```
Running Changeset: changelog.xml::my_example_1::jlyle [0.1 seconds]
Running Changeset: changelog.xml::my_example_2::jlyle [0.2 seconds]
Running Changeset: changelog.xml::my_example_3::jlyle [0.1 seconds]
```

RIght now I'm not sure if the current `Running Changeset: ...` is output before or after the actual execution of the changeset occurs. If it takes place after... well then it is easy to add the execution time on to the end! If that is output before the execution of the changeset occurs, then maybe print the text without a new line on the end, then add the execution time once the execution is completed and include the newline then.

Thoughts?

**Comment by MalloD12** [comment]
Hey Jason, 

Yeah, I meant in the execution output. Regarding where the "Running changeset..." log message it is logged before the changeset execution. See below:

```java
 Scope.getCurrentScope().getLog(getClass()).fine("Running Changeset: " + changeSet);
 fireWillRun(changeSet, databaseChangeLog, database, runStatus);
 ExecType execType;
 ObjectQuotingStrategy previousStr = this.database.getObjectQuotingStrategy();
  try {
       execType = changeSet.execute(databaseChangeLog, execListener, this.database);
```

Thanks,
Daniel.

**Comment by MalloD12** [comment]
@jasonlyle88 when you have a chance could you please resolve these merge conflicts that exist today and remove the part of adding a new column to DBCL table? I'll move this PR to a different column ("Candidate") in our board so we can prioritize the review of it when it's ready.

Thanks in advance,
Daniel.

**Comment by jasonlyle88** [comment]
@MalloD12 I won't be able to get to that reconciliation done today, but I can get to that in the next couple days for sure!

**Comment by MalloD12** [comment]
Oh yeah, I'm so sorry it wasn't my intention to hurry you up. I meant whenever you have some availability. We could be taking a loot at this next sprint or whenever it's ready.

Thank you, and have a great weekend!

Daniel.

**Comment by jasonlyle88** [comment]
@MalloD12 , Okay, the changes that affected the databasechangelog have been removed and the output to the console and to the log has been updated to include the changeset execution time (in milliseconds).

However, I chose not to put the execution time on the same line as the "Running Changeset: ..." line. This is because the changeset could have output (output change types, errors, plugins allowing output, etc). So I added just an additional line after everything has finished outputting Examle:

```
####################################################
##   _     _             _ _                      ##
##  | |   (_)           (_) |                     ##
##  | |    _  __ _ _   _ _| |__   __ _ ___  ___   ##
##  | |   | |/ _` | | | | | '_ \ / _` / __|/ _ \  ##
##  | |___| | (_| | |_| | | |_) | (_| \__ \  __/  ##
##  \_____/_|\__, |\__,_|_|_.__/ \__,_|___/\___|  ##
##              | |                               ##
##              |_|                               ##
##                                                ## 
##  Get documentation at docs.liquibase.com       ##
##  Get certified courses at learn.liquibase.com  ## 
##                                                ##
####################################################
Starting Liquibase at 19:26:43 using Java 21.0.4 (version [local build] #0 built at 2024-09-14 11:07+0000)
Liquibase Version: [local build]
WARNING: License service not loaded, cannot determine Liquibase Pro license status. Please consider re-installing Liquibase to include all dependencies. Continuing operation without Pro license.
Running Changeset: changelog.xml::test::jlyle
JMLTEST
JMLTEST
Changeset execution:90ms

UPDATE SUMMARY
Run:                          1
Previously run:               0
Filtered out:                 0
-------------------------------
Total change sets:            1

Liquibase: Update has been successful. Rows affected: 1
Liquibase command 'update' was executed successfully.

```

With this, this PR should be all ready to be reviewed

**Comment by MalloD12** [comment]
Awesome, thanks for updating this PR @jasonlyle88! I'll probably be looking at it today.

File: liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java [hunk 1]
```diff
@@ -962,15 +971,15 @@ public void rollback(Database database, ChangeExecListener listener) throws Roll
             if (runInTransaction) {
                 database.commit();
             }
+            stopInstant = Instant.now();
             rollbackExecType = ExecType.EXECUTED;
             Scope.getCurrentScope().addMdcValue(MdcKey.CHANGESET_OUTCOME, ExecType.EXECUTED.value.toLowerCase());
-            operationStopTime = new Date();
-            Scope.getCurrentScope().addMdcValue(MdcKey.CHANGESET_OPERATION_STOP_TIME, Instant.ofEpochMilli(operationStopTime.getTime()).toString());
+            Scope.getCurrentScope().addMdcValue(MdcKey.CHANGESET_OPERATION_STOP_TIME, stopInstant.toString());
             Scope.getCurrentScope().getLog(getClass()).fine("ChangeSet " + toString() + " has been successfully rolled back.");
```
Comment (MalloD12) [inline liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java:978]: Hey @jasonlyle88, don't you think we should be logging execution time here if the execution of the generated rollback statements were successfully executed?

**Review by MalloD12 — COMMENTED** [review]
Code changes look good to me. Thank you @jasonlyle88 for your patience and undestanding on all the disccussion around this change, and also for taking the time on updating the PR you already created some time ago. We very appreciate all your contributions to Liquibase. 

Regarding this change I have no major comments (other than the one I made for rollback execution), removing the part of updatring DBCL table, changes remains quite simple. I think we are ok and ready to move forward with this.

@rberezen I've manually checked this change and I think this is ok. As you might know we have an existent issue devops needs to fix and artifacts can't be generated right now,  but I think it wouldn't be a concern here since this is an OSS change that we don't have to validate against functional or test harness tests. What do you think?

File: liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java (hunk 1 above)
Comment (jasonlyle88) [reply liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java:978]: Good catch @MalloD12. I added a line to log the rollback execution time. Let me know if this works for you. If it does, feel free to resolve this conversation!

Comment (MalloD12) [reply liquibase-standard/src/main/java/liquibase/changelog/ChangeSet.java:978]: Nice, thanks @jasonlyle88!

**Review by MalloD12 — APPROVED** [review]

**Review by tati-qalified — CHANGES_REQUESTED** [review]
Hi @jasonlyle88, could you add some type of information on which changeset this message corresponds to? Either the id, or the same format used in "Running changeset ...", or something else.

Also just a small detail, could you add a space after `execution:`?

Thank you!

**Comment by tati-qalified** [comment]
@jasonlyle88 that looks great, thank you for your contribution!

**Review by tati-qalified — APPROVED** [review]

**Review by filipelautert — APPROVED** [review]

**Comment by MalloD12** [comment]
Hi @jasonlyle88,

Our QA team was testing this change yesterday and made us notice there are some additional log messages introduced with this PR that would become logs much more verbose for big changelogs. These messages are:

```console
Ran Changeset : changelog.xml::upperCaseTable::Liquibase Pro User [4ms]
[2024-10-02 12:53:48] INFO [liquibase.ui] Ran Changeset  : changelog.xml::upperCaseTable::Liquibase Pro User [4ms]
``` 

These additional log messages are the ones we are introducing from `UpdateVisitor`. Could you please remove those? Or if there is any specific reason I don't remember now could you please let me know so I can discuss it with the team?

Thanks,
Daniel.

**Comment by jasonlyle88** [comment]
Hey @MalloD12 ,

I run have the following code for logging the `Ran Changeset` lines:

```java
if (!(executor instanceof LoggingExecutor)) {
    Scope.getCurrentScope().getUI().sendMessage("Ran Changeset    : " + changeSet + " [" + changeSet.getExecutionMilliseconds() + "ms]");
}
Scope.getCurrentScope().getLog(getClass()).fine("Ran Changeset    : " + changeSet + " [" + changeSet.getExecutionMilliseconds() + "ms]");
```

This is done twice, because I mimiced what you all did for the `Running Changeset` logging:

```java
if (!(executor instanceof LoggingExecutor)) {
    Scope.getCurrentScope().getUI().sendMessage("Running Changeset: " + changeSet);
}
Scope.getCurrentScope().getLog(getClass()).fine("Running Changeset: " + changeSet);
```

Is this what you are referring to?

**Comment by MalloD12** [comment]
Hey Jason, 

Correct, that's what I'm talking about. We think that only logging:

```java
log.info("ChangeSet " + toString(false) + " ran successfully in " + (stopTime - startTime) + "ms");
```  

Is good enough. 

The team was highlighting the other day a comparison between `4.29.2` and changes from this branch:

`4.29.2`:
```console
Running Changeset: changelog.xml::upperCaseTable::Liquibase Pro User
[2024-10-02 12:51:14] INFO [liquibase.ui] Running Changeset: changelog.xml::upperCaseTable::Liquibase Pro User
[2024-10-02 12:51:14] INFO [liquibase.changelog] Table UPPERCASETHIS created
[2024-10-02 12:51:14] INFO [liquibase.changelog] ChangeSet changelog.xml::upperCaseTable::Liquibase Pro User ran successfully in 4ms
```

Changes from this branch:
```console
Running Changeset: changelog.xml::upperCaseTable::Liquibase Pro User
[2024-10-02 12:53:48] INFO [liquibase.ui] Running Changeset: changelog.xml::upperCaseTable::Liquibase Pro User
[2024-10-02 12:53:48] INFO [liquibase.changelog] Table UPPERCASETHIS created
[2024-10-02 12:53:48] INFO [liquibase.changelog] ChangeSet changelog.xml::upperCaseTable::Liquibase Pro User ran successfully in 4ms
Ran Changeset    : changelog.xml::upperCaseTable::Liquibase Pro User [4ms]
[2024-10-02 12:53:48] INFO [liquibase.ui] Ran Changeset    : changelog.xml::upperCaseTable::Liquibase Pro User [4ms]
```

After a team discussion we had on Wednesday, we agreed on just logging what it was before, so it would require removing those new lines added in `UpdateVisitor` class.

Thanks,
Daniel.

**Comment by jasonlyle88** [comment]
I'm confused. Those two lines in the `UpdateVisitor` class are the only place that the execution time is getting logged. If I remove that, it isn't getting logged anywhere. If I remove those lines, the execution time isn't logged anywhere... which is the purpose of this PR.

The line you provided:

```java
log.info("ChangeSet " + toString(false) + " ran successfully in " + (stopTime - startTime) + "ms");
```

Does not exist anywhere. There is also no `log` variable that exists in the `UpdateVisitor` where I currently have my logging lines, so I don't think thats something you wanted me to add there.

Can you please go into more detail about what you decided? Or make the change as your team expects it to run and I can then review it?

**Comment by jasonlyle88** [comment]
Hey @MalloD12 ,

Okay, I see that now in the `ChangeSet` class, I was looking in the `UpdateVisitor`. Thanks, I didn't realize this was already being logged at the info level 🤦🏼‍♂️.

I think it would be nice to be able to see execution time without having to see everything else that comes at the info log level though, which greatly increases the output size.

What would the team thing about a property that determines if the execution time is output to the console in the format setup in this PR? Obviously defaulted to off, but that teams could turn on to see execution timing without having to turn on info logging?

If the team is open to a preference to do this... I would also appreciate some guidance in this area. I've never looked into that side of the project, so any pointers on where to begin would be appreciated!

**Comment by MalloD12** [comment]
Hey @jasonlyle88,

Let me check that with the team. In case they agree to add a new property to isolate this logging, I think it would probably be better to separate that change into a new issue, so even though things are related, we don't mix many things/discussions here. I'll let you know what they say about it.

Thanks,
Daniel.

**Comment by jasonlyle88** [comment]
I was thinking the new property would only control if the "Ran changeset" lines that I currently have codeded in this PR get output or not. Something like the following:

```java
if (getLiquibaseProperty('displayConsoleExecutionTime')) {
    if (!(executor instanceof LoggingExecutor)) {
        Scope.getCurrentScope().getUI().sendMessage("Ran Changeset    : " + changeSet + " [" + changeSet.getExecutionMilliseconds() + "ms]");
    }
    Scope.getCurrentScope().getLog(getClass()).fine("Ran Changeset    : " + changeSet + " [" + changeSet.getExecutionMilliseconds() + "ms]");
}
```

Where `getLiquibaseProperty(displayConsoleExecutionTime)` is just my pseudocode for how to get the new property since I don't currently know how any of that works

**Comment by MalloD12** [comment]
Hey @jasonlyle88,

Team said they would prefer to leave this PR as it's to not add some extra logic for this.  If there are additional changes it would probably be worth writing a dedicated extension for it, but we think we should leave this as it is, and move forward with it. 

Hope you can understand this. Thanks a lot for constantly improving Liquibase and always adding new features to make this a better project.

Thanks,
Daniel.

**Review by rberezen — APPROVED** [review]

**Comment by filipelautert** [comment]
@MalloD12 could you resolve the conflicts so we can merge it?

**Comment by MalloD12** [comment]
Hey @jasonlyle88 ,

Could you please push the latest changes you made for this PR again? Two days ago, I resolved some merge conflicts, and by mistake, I included the `EXECUTION_MILLISECONDS` column, which seems to have overridden some of your other commits.  If you still have this available on your local, please push them and create a new PR for it.

Thanks and apologies for this inconvenient 🙏,
Daniel.

**Comment by jasonlyle88** [comment]
@MalloD12 I believe I've done what you want in #6565

**Pull request merged**